File: Toolkit/Data_Structures/DisjointSet.cpp
#include <vector>

using namespace std;
class DisjointSet {
      public:
      vector<int> depth;
      vector<int> parents;
      int N;
      DisjointSet(int n) {
         N = n;
         for (int i = 0; i < N; i++) {
            depth.push_back(0);
            parents.push_back(i);
         }
      }
      int findSet(int x) {
         if (parents[x] != x) {
            parents[x] = findSet(parents[x]);
         }
         return parents[x];
      }
      void unionSets(int x, int y) {
         int xRep = findSet(x);
         int yRep = findSet(y);
         if (xRep == yRep) {
            return;
         }
         if (depth[xRep] > depth[yRep]) {
            parents[yRep] = xRep;
         } else if (depth[yRep] > depth[xRep]) {
            parents[xRep] = yRep;
         } else {
            depth[xRep] += 1;
            parents[yRep] = xRep;
         }
      }
};

File: Templates/Data_Structures/DisjointSet.java
class DisjointSet
{
    int[] depth, parent;
    int N;

    public DisjointSet(int N)
    {
        depth = new int[N];
        parent = new int[N];
        this.N = N;
        makeSet();
    }

    void makeSet()
    {
        for (int i = 0; i < N; i++)
        {
            parent[i] = i;
        }
    }

    int findSet(int x)
    {
        if (parent[x] != x)
        {
            parent[x] = findSet(parent[x]);
        }
        return parent[x];
    }

    void union(int x, int y)
    {
        int xRep = findSet(x), yRep = findSet(y);

        if (xRep == yRep)
        {
            return;
        }

        if (depth[xRep] < depth[yRep])
        {
            parent[xRep] = yRep;
        } else if (depth[yRep] < depth[xRep])
        {
            parent[yRep] = xRep;
        } else {
            parent[yRep] = xRep;
            depth[xRep] = depth[xRep] + 1;
        }
    }
}
